azerothcore/azerothcore-wotlk
# Fix(Scripts/BRD) - Magmus, Emperor, Priestess, Arena improvements

## Changes Proposed:
Apologies for the huge PR, will try to figure out a way to make following ones smaller.

- Updated Moira Bronzebeard to bossAI. She now casts all her spells and emotes
- Updated Moira Bronzebeard. She is replaced by High priestess of Thaurissan if everybody has saved the princess in a previous run. High Priestess uses her yell. 
- Added High Priestess yell to database
- Updated emperor to BossAI.
- Emperor no longer casts hand of thaurissan at victim but randomly
- Updated emperor to do his yells.
- After arena, spectators become neutral (please verify faction I used)
- Arena bosses are now actually random. Before we were using rand() without a seed and you seemingly always got small spiders and big spider boss
- Magmus is now casting his warstomp from 100% hp. Rewrote to BossAI. ~Correctly yells after opening doors~ was outside.
- Fire statues at Magmus now start on engage. They have random grouping and more random in their cast.
- Flame keepers are now pooled. Only maximum 2 of them per dungeon at any time

Edit with additional fixes:
- Added scripts for arena bosses
- Priestess and Moira are now mages and not paladins.
- reworked the Lyceum event a bit too, since the data for the dungeon was not set properly. Doors are opened by the instance and not by the braziers anymore (fundamentally, it's the same for the players, but I needed setData to work for Moira/Priestess swap which is done after the lyceum event to have Moira spawned).

## Issues Addressed:

- closes https://github.com/azerothcore/azerothcore-wotlk/issues/2689
- closes https://github.com/azerothcore/azerothcore-wotlk/issues/1774 (was mainly closed already, now just some cosmetics and logic)
- closes https://github.com/azerothcore/azerothcore-wotlk/issues/1779
- closes https://github.com/azerothcore/azerothcore-wotlk/issues/7726

## SOURCE:

- flamekeepers: https://wowpedia.fandom.com/wiki/Lyceum
- other stuff can be seen here in classic https://youtu.be/hnRIUMTjygU?t=12035 but is common sense
- Priestess yell https://wowpedia.fandom.com/wiki/High_Priestess_of_Thaurissan 

## Tests Performed:

- builds on my machine with VS 2019 on win10.
- poked around on fights


## How to Test the Changes:


0. remove GM tag before entering, or you will not get the priestess
1. Go to arena, you should be able to see something else than borers and a big blue spider at the end.
2. Finish arena, mobs upstairs should be neutral.
3. Go to lyceum, should be only 2 flame keepers
4. Fire statues should be off until engaging Magmus
5. Go to emperor room and start killing mobs, yells should happen. Pull the boss, he should have correct yells. It's good to have several people here to confirm that the random cast is done correctly.
6. Finish quest 4004 or 4363 and start again (reset dungeon), Moira should be gone and the priestess took her place. Note that this happens when the doors to Magmus open, so you need to finish the lyceum event.

7. Try all arena bosses 1 by 1 (probably just .add them) to verify the spells

## Known Issues and TODO List:


- [ ] Spells need to be defined before merging. Valid for all arena bosses and Moira/Priestess
- [ ] Arena waves need to be confirmed, how many mobs of each should spawn.
- [ ] Timers need to be defined
- [ ] The spawnpoints and waypoints of the flame keepers need to be reviewed and changed. Should be 8 keepers
- [ ] If Emperor has done 5 yells for the senators (ie all of them), he will not yell "Come to aid the throne" if a mob is still alive in the room. This is an edge case that I couldn't find an easy "GetCreaturesWithinRange()" to fix so I didn't bother.


## Credit
Thanks to https://github.com/Asthalor for the sql request for the yells for Emperor.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
